Check protocol version compatibility during a platform update

gosusnp opened the pull request:
## What

When updating the platform we need to ensure that we are not going to end up having to disable a connection because the support range of the protocol version changed.
This PR adds the logic to check for protocol version compatibility of the platform and the connectors used in active syncs to decide whether it is safe to continue the upgrade or if we need to other interventions before.

Relates to #15463 

## How
* Add a ProtocolVersionChecker to check the protocol compatibilty
* Plug it in the Bootloader before the DB migrations in case we want to abort the upgrade
* Small refactoring for DB code reuse

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨

Following this change, platform upgrades can be aborted because of protocol version conflicts.
If such cases happen, there will need to be some manual operations before being able to upgrade the platform. A typical operation would be upgrading a connector to a version that is using a more recent version of the protocol.

jdpgrailsdev commented on airbyte-bootloader/src/main/java/io/airbyte/bootloader/BootloaderApp.java:
```diff
@@ -249,7 +291,11 @@ public static void main(final String[] args) throws Exception {
+
+      final var bootloader =
+          new BootloaderApp(configs, featureFlags, secretMigrator, configsDslContext, jobsDslContext, configsFlyway, jobsFlyway, definitionsProvider,
+              false);
```
I assume that this is hard-coded to `false` just for now and will eventually be replaced by some sort of configuration property?

gosusnp replied:
Yes, I had the change to add the configs, however it felt more misleading for the time being because this how we apply upgrades is currently hardcoded and injected in cloud.
The config should control both how apply the upgrades and how we validate.
Unless someone has a strong feeling we should address this now, I'd rather push it to when we micronaut the bootloader.

Should I move [this comment](https://github.com/airbytehq/airbyte/pull/19200/files#diff-887184626b5be3420f725a31a95dfde933c7e85f678d7f8de66ea6caf86462d4R82) to here instead?

jdpgrailsdev commented on airbyte-bootloader/src/main/java/io/airbyte/bootloader/ProtocolVersionChecker.java:
```diff
@@ -0,0 +1,218 @@
+    final Map<ActorType, Set<UUID>> conflicts = getConflictingActorDefinitions(targetRange);
+
+    if (conflicts.isEmpty()) {
+      log.info("No conflicts");
```
Nit:  Consider adding more context to this message so it's easy to understand what has no conflicts.  Perhaps something like "No conflicting connector protocol versions detected."

gosusnp replied:
updated

jdpgrailsdev commented on airbyte-bootloader/src/main/java/io/airbyte/bootloader/ProtocolVersionChecker.java:
```diff
@@ -0,0 +1,218 @@
+      // hard.
+      // If the new config is fine, the system should self-heal.
+      log.warn("Inconsistent AirbyteProtocolVersion found, only one of min/max was found. (min:{}, max:{})",
+          min.map(Version::serialize).orElse(""), max.map(Version::serialize).orElse(""));
```
Nit:  consider extracting to a function that can be reused for both `min` and `max`.  Something like:

```
private Function<Optional<Version>, String> versionFunction = v -> v.map(Version::serialize).orElse("");

...

log.warn("... (min:{}, max:{})", versionFunction.apply(min), versionFunction.apply(max));
```

It doesn't reduce the code that much, but does prevent repetition of the logic to extract the version from the optional to avoid potential drift.

gosusnp replied:
Yes, this makes sense, but I am thinking of pushing it a bit further and changing the `toString()` implementation and stop using `serialize` everywhere.

The different between the two is about format, `serialize()` gives a `<major>.<minor>.<patch>` while `toString()` returns `Version{version='<major>.<minor>.<patch>', major='<major>', minor='<minor>', patch='<patch>'}`.

jdpgrailsdev commented on airbyte-bootloader/src/main/java/io/airbyte/bootloader/ProtocolVersionChecker.java:
```diff
@@ -0,0 +1,218 @@
+            return String.format("Destination: %s: %s: protocol version: %s",
+                destDef.getDestinationDefinitionId(), destDef.getName(), destDef.getProtocolVersion());
+          } catch (Exception e) {
+            log.info("Failed to getStandardDestinationDefinition for " + defId, e);
```
Nit:  use placeholders instead of string concatenation in log message.

gosusnp replied:
done

benmoriceau commented on airbyte-bootloader/src/main/java/io/airbyte/bootloader/ProtocolVersionChecker.java:
```diff
@@ -0,0 +1,218 @@
+    if (actorType == ActorType.SOURCE) {
+      stream = definitionsProvider.getSourceDefinitions()
+          .stream()
+          .map(def -> Map.entry(def.getSourceDefinitionId(), AirbyteProtocolVersion.getWithDefault(def.getSpec().getProtocolVersion())));
```
Why not doing collect(Collectors.toMap) here? or just map the to get the keys. It looks like this is the only thing that is neeed here.

benmoriceau commented on airbyte-bootloader/src/main/java/io/airbyte/bootloader/ProtocolVersionChecker.java:
```diff
@@ -0,0 +1,218 @@
+            return String.format("Source: %s: %s: protocol version: %s",
+                sourceDef.getSourceDefinitionId(), sourceDef.getName(), sourceDef.getProtocolVersion());
+          } catch (Exception e) {
+            log.info("Failed to getStandardSourceDefinition for {}", defId, e);
```
Should this be an error?

benmoriceau commented on airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/ConfigWriter.java:
```diff
@@ -29,6 +41,36 @@
+   */
+  static Set<String> getConnectorRepositoriesInUse(final DSLContext ctx) {
+    return getActorDefinitionsInUse(ctx)
+        .map(Record4::value2)
```
can we use `map(record -> record.get(ACTOR_DEFINITION.DOCKER_REPOSITORY))` here? This would help to understand what we are fetching here. (Same bellow)

gosusnp replied:
Much better indeed. done.

gosusnp commented:
@jdpgrailsdev, @benmoriceau, FYI, I just changed the interface a bit to accommodate cloud behavior.

With the addition of the `RemoteDefinitionsProvider`, the `DefinitionsProvider` is now optional. Currently, if we fail to initialize a valid `RemoteDefinitionsProvider`, we continue to start the platform without the ability to update connectors from seed. I have updated the logic to make sure that we do not prevent the platform from starting if there is no conflicts detected even without a `DefinitionsProvider`.
